permissioned-token/program/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+            mint.key,
+            Rent::get()?.minimum_balance(space),
+            space as u64,
+            token_program.key,
+        ),
+        &[payer.clone(), mint.clone(), system_program.clone()],
+    )?;
+    let (mint_authority, _) = Pubkey::find_program_address(&[mint.key.as_ref()], program_id);
+    let (freeze_key, _) = get_freeze_authority(upstream_authority.key);
+    initialize_mint(&freeze_key, &mint_authority, mint, token_program, decimals)
+}
+
+pub fn process_initialize_account(accounts: &[AccountInfo]) -> ProgramResult {
+    let InitializeAccount {
+        token_account,
+        owner,
+        payer,
+        upstream_authority,
+        freeze_authority,
+        mint,
+        system_program,
+        rent,
+        associated_token_program,
+        token_program,
+    } = InitializeAccount::load(accounts)?;
+    invoke(
+        &create_associated_token_account(payer.key, owner.key, mint.key, token_program.key),
+        &[
+            associated_token_program.clone(),
+            payer.clone(),
+            owner.clone(),
+            token_account.clone(),
+            mint.clone(),
+            system_program.clone(),
+            token_program.clone(),
+            rent.clone(),
+        ],
+    )?;
+    let seeds = get_freeze_authority_seeds_checked(upstream_authority.key, freeze_authority.key)?;
+    freeze(freeze_authority, mint, token_account, token_program, &seeds)
+}
+
+pub fn process_transfer(accounts: &[AccountInfo], amount: u64) -> ProgramResult {
+    let Transfer {
+        src_account,
+        dst_account,
+        mint,
+        owner,
+        upstream_authority,
+        freeze_authority,
+        token_program,
+    } = Transfer::load(accounts)?;
+    let seeds = get_freeze_authority_seeds_checked(upstream_authority.key, freeze_authority.key)?;
+    thaw(freeze_authority, mint, src_account, token_program, &seeds)?;
+    thaw(freeze_authority, mint, dst_account, token_program, &seeds)?;
+    transfer(src_account, dst_account, owner, token_program, amount)?;
+    freeze(freeze_authority, mint, dst_account, token_program, &seeds)?;
+    freeze(freeze_authority, mint, src_account, token_program, &seeds)
+}
+
+pub fn process_mint_to(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    amount: u64,
+) -> ProgramResult {
+    let MintOrBurn {
+        mint,
+        token_account,
+        owner,
+        upstream_authority,
+        freeze_authority,
+        token_program,
+    } = MintOrBurn::load(accounts)?;
+    let seeds = get_freeze_authority_seeds_checked(upstream_authority.key, freeze_authority.key)?;
+    thaw(freeze_authority, mint, token_account, token_program, &seeds)?;
+    let (mint_authority, bump) = Pubkey::find_program_address(&[mint.key.as_ref()], program_id);
+    assert_with_msg(
+        &mint_authority == owner.key,
+        spl_token::error::TokenError::OwnerMismatch,
+        "Mint authority must be a PDA of the Permissioned Token Program",
+    )?;
```

Comment:
This check confused me for awhile until I realized that `owner` is the mint authority in this context.  Can it be removed?

Reply:
Didn’t want to create 2 separate account context objects

Reply:
But I can alias the variable